### ISSUE TITLE
Data descriptor type is changed after assignment

### ISSUE THREAD

**Issue by ermishechkin** [microsoft/pyright#1814]
**Describe the bug**
Type of instance member is changed after assignment even if it is data descriptor and contains \_\_set\_\_ method. In such cases return type of \_\_get\_\_ method is more correct.

**Expected behavior**
The same as mypy (see code below):
note: Revealed type is 'builtins.int'
note: Revealed type is 'builtins.int'

**Screenshots or Code**
```python
from typing import Any


class AlwaysIntMember:
    def __get__(self, instance: Any, owner: Any) -> int:
        ...

    def __set__(self, instance: Any, value: str) -> None:
        ...

class Struct:
    member = AlwaysIntMember()

inst = Struct()
reveal_type(inst.member)  # Type of "inst.member" is "int"
inst.member.bit_length()  # OK
inst.member = '123'
reveal_type(inst.member)  # Type of "inst.member" is "Literal['123']"
inst.member.bit_length()  # error: Cannot access member "bit_length" for type "Literal['123']"
                          # error: Type of "bit_length" is unknown (reportUnknownMemberType)
                          # info: Type of "inst.member.bit_length()" is "Unknown"
```

**VS Code extension or command-line**
pyright@1.1.136

**Comment by erictraut** [comment]
Pyright uses type narrowing for assignment expressions. It assumes that if you assign a value to a member, you will receive the same value back when you access the member within the same execution scope. This assumption holds true most of the time, and this type narrowing is very useful. You've found an example where that assumption breaks down because you're implementing a descriptor that is asymmetrical (i.e. the setter accepts different types than the getter). While Python allows this, I'd argue that it's not a good design practice. In this case, I would personally avoid a descriptor class and instead use an explicit `set_x` method to make it clear to the caller that the code is performing a runtime type conversion.

Let's try to understand why mypy doesn't report any errors in this case. After all, mypy also performs type narrowing for assignment expressions and generally makes the same assumption about member accesses. It appears that mypy includes some special case logic where it skips type narrowing in certain cases, including the one in your code sample. Special cases like this in a type checker run the risk of fixing some cases and breaking others. That appears to be the case with mypy's heuristic here.

Consider the following example:
```python
from typing import Any, Protocol, Union

class HasName(Protocol):
    def __name__(self) -> str:
        ...

class Animal:
    def __name__(self) -> str:
        ...

class Dog(Animal):
    def __name__(self) -> str:
        ...

    def run(self) -> None:
        print("Run Spot Run!")

    def bark(self) -> None:
        print("Ruff!")

class Cat:
    def __name__(self) -> str:
        ...

    def run(self) -> None:
        print("Run Kitty Run!")

    def purr(self) -> None:
        print("Purr!")

class AnimalMember:
    def __get__(self, instance: Any, owner: Any) -> Animal:
        ...

    def __set__(self, instance: Any, value: HasName) -> None:
        ...

class Struct:
    member = AnimalMember()

def test_descriptor(cat: Cat, dog: Dog, cat_or_dog: Union[Cat, Dog]):
    inst = Struct()

    inst.member = cat_or_dog
    reveal_type(inst.member)  # mypy: Revealed type is 'Animal'
    inst.member.run()  # mypy: "Animal" has no attribute "run"
    inst.member.bark()  # mypy: "Animal" has no attribute "bark"
    inst.member.purr()  # mypy: "Animal" has no attribute "purr"

    inst.member = cat
    reveal_type(inst.member)  # mypy: Revealed type is 'Animal'
    inst.member.run()  # mypy: "Animal" has no attribute "run"
    inst.member.bark()  # mypy: "Animal" has no attribute "bark"
    inst.member.purr()  # mypy: "Animal" has no attribute "purr"

    inst.member = dog
    reveal_type(inst.member)  # mypy: Revealed type is 'Dog'
    inst.member.run()  # mypy: no error
    inst.member.bark()  # mypy: no error
    inst.member.purr()  # mypy: "Dog" has no attribute "purr"

    inst.member = cat_or_dog
    reveal_type(inst.member)  # mypy: Revealed type is 'Dog' !!!!
    inst.member.run()  # mypy: no error
    inst.member.bark()  # mypy: no error
    inst.member.purr()  # mypy: "Dog" has no attribute "purr"

    inst.member = cat
    reveal_type(inst.member)  # mypy: Revealed type is 'Dog' !!!!
    inst.member.run()  # mypy: no error
    inst.member.bark()  # mypy: no error !!!!
    inst.member.purr()  # mypy: "Dog" has no attribute "purr" !!!!
```

From what I can tell, mypy is applying assignment-based type narrowing for descriptors in the case that the assigned type is a subtype of the getter's return type and skipping type narrowing in other cases. As you can see from the example above, that heuristic isn't sound and breaks down in some cases.

Arguably, a better heuristic is to skip assignment-based type narrowing in the case where the following hold true:
1. The target member is a descriptor (or property)
2. There is both a getter and a setter and neither of them have function overloads
3. The setter's value parameter type and the getter's return type are both explicitly annotated
4. The setter's value parameter type is not a subtype of the getter's return type

As I said, I'm generally reluctant to add special cases and heuristics unless I'm confident that they'll do more good than harm. They tend to complicate the type checking logic and make it hard for users to understand what the type checker is doing in different situations.

I'll need to think about it a bit more before I convince myself that this special-case logic is worth it in this case.

**Comment by erictraut** [comment]
I think that my proposed heuristic above will accomplish what you're suggesting. Mypy's current behavior is clearly wrong (see my example above, notably the lines with "!!!!").

**Comment by ermishechkin** [comment]
Why is there condition about overloads in the proposed heuristic? It limits common case - owner-based __get__ overloading (Literal[None] or Type[T]).

**Comment by erictraut** [comment]
If there are overloads, it's not generally possible to evaluate point #4 ("The setter's value parameter type is not a subtype of the getter's return type") because there can be multiple setter signatures and multiple getter signatures. Do you see a way around this problem?

My assumption is that performing type narrowing on assignment should be the default behavior (since it's generally safe and useful), so I'm looking for a minimal safe heuristic for when to avoid this default behavior.

**Comment by ermishechkin** [comment]
May be it is possible to combine described solutions: treat get/set access to descriptor as function call and use type narrowing.

For assignment:
1a. check new value type based on type declared in \_\_set\_\_ method
1b. save type of assigned value

For get access:
2a. deduce return type of \_\_get\_\_ method (it supports overloading)
2b. if previously saved type (1b) is subtype of deduced type (2a) then return previously saved type (it supports type narrowing). else return deduced type (2a)

**Comment by erictraut** [comment]
I think the heuristic I outlined above is the correct approach, but this change will involve adding expensive checks to a hot code path in the analyzer. Since this is such an edge case and involves a use case that I think is unadvisable, I don't think it's worth slowing down analysis performance for everyone to accommodate this. We might reconsider in the future if there is sufficient interest, but for now, I'm going to close this as "won't fix".

**Comment by wkrasnicki** [comment]
I'm not sure I understand why it was decided to leave this behaviour, as from what I tested descriptors are currently unusable in Pyright, which is a shame, because we rely heavily on them in our code.

A proper descriptor type hint requires an overload, because we need to handle the class access case, which returns a different type than when we access the descriptor on an instance. This is not handled by Pyright, because of rule 2 in your heuristic. While this case is a rare occurance, so we could work with incorrect types, the narrowing still happens when there is no overload!

```
class PositiveInt:
    def __set_name__(self, owner: type[object], name: str) -> None:
        self.name = name

    def __get__(
        self, obj: object, objtype: type[object] | None = None
    ) -> int:
        return obj.__dict__[self.name]

    def __set__(self, obj: object, value: int) -> None:
        if value <= 0:
            raise ValueError(f"{self.name} must be positive")
        obj.__dict__[self.name] = value

    def __delete__(self, obj: object) -> None:
        del obj.__dict__[self.name]

class Widget:
    rotations = PositiveInt()


widget = Widget()
widget.rotations = 12
widget.rotations # The type has been narrowed to Literal[12]
```

Is there something I misunderstood about your heuristic?
